packages/@aws-cdk/aws-ecs/lib/base/base-service.ts

```diff
@@ -0,0 +1,283 @@
+
+  /**
+   * Time after startup to ignore unhealthy load balancer checks.
+   *
+   * @default ??? FIXME
+   */
+  healthCheckGracePeriodSeconds?: number;
+
+  /**
+   * Fargate platform version to run this service on
+   *
+   * Unless you have specific compatibility requirements, you don't need to
+   * specify this.
+   *
+   * @default Latest
+   */
+  platformVersion?: FargatePlatformVersion;
+}
+
+/**
+ * Base class for Ecs and Fargate services
+ */
+export abstract class BaseService extends cdk.Construct
+  implements elbv2.IApplicationLoadBalancerTarget, elbv2.INetworkLoadBalancerTarget, cdk.IDependable {
+
+  /**
+   * CloudFormation resources generated by this service
+   */
+  public readonly dependencyElements: cdk.IDependable[];
+
+  /**
+   * Manage allowed network traffic for this construct
+   */
+  public abstract readonly connections: ec2.Connections;
+
+  /**
+   * ARN of this service
+   */
+  public readonly serviceArn: string;
+
+  /**
+   * Name of this service
+   */
+  public readonly serviceName: string;
+
+  /**
+   * Name of this service's cluster
+   */
+  public readonly clusterName: string;
+
+  protected loadBalancers = new Array<cloudformation.ServiceResource.LoadBalancerProperty>();
+  protected networkConfiguration?: cloudformation.ServiceResource.NetworkConfigurationProperty;
+  protected readonly abstract taskDef: BaseTaskDefinition;
+  protected _securityGroup?: ec2.SecurityGroupRef;
+  private readonly resource: cloudformation.ServiceResource;
+  private scalableTaskCount?: ScalableTaskCount;
+
+  constructor(parent: cdk.Construct, name: string, props: BaseServiceProps, additionalProps: any, clusterName: string) {
+    super(parent, name);
+
+    this.resource = new cloudformation.ServiceResource(this, "Service", {
+      desiredCount: props.desiredCount || 1,
+      serviceName: props.serviceName,
+      loadBalancers: new cdk.Token(() => this.loadBalancers),
+      deploymentConfiguration: {
+        maximumPercent: props.maximumPercent || 200,
+        minimumHealthyPercent: props.minimumHealthyPercent || 50
+      },
+      /* role: never specified, supplanted by Service Linked Role */
+      networkConfiguration: new cdk.Token(() => this.networkConfiguration),
+      platformVersion: props.platformVersion,
+      ...additionalProps
+    });
+    this.serviceArn = this.resource.serviceArn;
+    this.serviceName = this.resource.serviceName;
+    this.dependencyElements = [this.resource];
+    this.clusterName = clusterName;
+  }
+
+  /**
+   * Called when the service is attached to an ALB
+   *
+   * Don't call this function directly. Instead, call listener.addTarget()
+   * to add this service to a load balancer.
+   */
+  public attachToApplicationTargetGroup(targetGroup: elbv2.ApplicationTargetGroup): elbv2.LoadBalancerTargetProps {
+    const ret = this.attachToELBv2(targetGroup);
+
+    // Open up security groups. For dynamic port mapping, we won't know the port range
```

Comment:
Is this the best practice? Make sure we document this somehere

Reply:
It's the only practice given this configuration. The host is going to pick random ports from the ephemeral range for every task, so this is the only thing we can do to make it work.

Reply:
Got it. Maybe worth adding a comment with a word or two